pkg/usagestats/reporter.go

```diff
@@ -109,8 +109,20 @@ func (r *Reporter) running(ctx context.Context) error {
 
 	level.Info(r.logger).Log("msg", "usage stats reporter initialized", "cluster_id", seed.UID)
 
-	// Find when to send the next report. We want all instances of the same Mimir cluster computing the same value.
-	nextReportAt := nextReport(r.reportSendInterval, seed.CreatedAt, time.Now())
+	// Keep track of the next report to send, so that we reuse the same on retries after send failures.
+	var nextReport *Report
+	var nextReportAt time.Time
+
+	// We define a function to update the timestamp of the next report to make sure
+	// we also reset the next report when doing it.
+	updateNextReportAt := func() {
```

Comment:
Opinionated: I think this function's name is hiding the side effect of upsating `nextReport` and makes the code below harder to read than if we repeated the two lines in both places.

Reply:
I still think the function may help avoiding to forget to reset the `nextReport` to `nil` in future code changes. I agree `updateNextReportAt()` naming doesn't clarify it, so I've renamed it to `scheduleNextReport()`.